Python/chapter02/2.3 - Delete Middle Node/miguel_2.3_soln.py

```diff
@@ -0,0 +1,161 @@
+        linked list looks like
+        a -> b -> d -> e -> f
+"""
+import unittest
+
+
+class Node:
+    def __init__(self, d: int):
+        self.data = d
+        self.next = None
+
+    def __repr__(self):
+        return self.__str__()
+
+    def __str__(self):
+        return '<Node Value: {}>'.format(self.data)
+
+    def __eq__(self, other: object):
+        if not isinstance(other, Node):
+            return NotImplemented
+        return self.data == other.data
+
+
+class LinkedList:
+    def __init__(self, *numbers: int):
+        self.head = None
+        self.tail = None
+        self.size = 0
+        for num in numbers:
+            self.append_to_tail(num)
+
+    def append_to_tail(self, d: int) -> None:
+        if self.head is None:
+            self.head = Node(d)
+            self.tail = self.head
+        else:
+            end = Node(d)
+            self.tail.next = end
+            self.tail = end
+        self.size += 1
+
+    def append_to_head(self, d: int) -> None:
+        new_head = Node(d)
+        new_head.next = self.head
+        self.head = new_head
+        self.size += 1
+
+    def get_node_at(self, index: int) -> Node:
+        if index < 0 or index >= self.size:
+            raise IndexError('list index out of range')
+        n = self.head
+        i = 0
+        while n is not None:
+            if i == index:
+                return n
+            i += 1
+            n = n.next
+
+    def __repr__(self):
+        return self.__str__()
+
+    def __str__(self):
+        if self.head is None:
+            return '<empty>'
+        ll = []
+        n = self.head
+        while n.next is not None:
+            ll.append('{} -> '.format(n.data))
+            n = n.next
+        ll.append(str(n.data))
+        return ''.join(ll)
+
+    def __eq__(self, other: object):
+        if not isinstance(other, LinkedList):
+            return NotImplemented
+        a = self.head
+        b = other.head
+        while a is not None and b is not None:
+            if a.data != b.data:
+                return False
+            # otherwise, advance both pointers
+            a = a.next
+            b = b.next
+        return a is None and b is None
+
+
+def delete_middle_node(ll: LinkedList, node: Node) -> LinkedList:
+    """
+    delete_middle_node will delete a node from
+    a singly linked list. The node can be any
+    node within the linked list that is NOT the
+    head or the tail.
+    We will be comparing the address of
+    the input node with the address of each
+    node in the linked list.
+    Runtime: O(N)
+    Space Complexity: O(1)
+    :param ll: an input linked list
+    :param node: pointer to a node in ll
+    :return: a linked list with/wo having deleted node
+    """
+    if node is ll.head or node is ll.tail:
+        raise ValueError('node cannot be head or tail of linked list')
+    n = ll.head
+    while n.next is not None:
+        if n.next is node:
+            n.next = n.next.next
```

Comment:
here you might as well return, since the way you create the linked list, there can be at most one reference to the specified node (although in degenerate cases, linked lists can have cycles in them).